routing/index_graph.cpp

```diff
@@ -1,9 +1,94 @@
+  {
+    if (featureIdFrom != featureIdTo)
+      return true;
+
+    // @TODO(bykoianko) According to current code if a feature id is marked as a feature with
+    // restrictricted U-turn it's restricted to make a U-turn on the both ends of the feature.
+    // Generally speaking it's wrong. In osm there's information about the end of the feature
+    // where the U-turn is restricted. It's necessary to pass the data to mwm and to use it here.
+    // Please see test LineGraph_RestrictionF1F1No for details.
+    return u.GetSegmentIdx() == v.GetSegmentIdx() && u.IsForward() != v.IsForward();
+  }
+
+  // Taking into account that number of restrictions of type Only starting and
+  // ending with the same feature id is relevantly small it's possible to ignore such cases.
+  if (featureIdFrom == featureIdTo)
+    return false;
+
+  // Looking for a range of restrictins of type Only starting from |featureIdFrom|
+  // and finishing with any feature.
+  auto const range = equal_range(restrictions.cbegin(), restrictions.cend(),
+                                 Restriction(Restriction::Type::Only, {featureIdFrom}),
+                                 [](Restriction const & r1, Restriction const & r2) {
+                                   CHECK(!r1.m_featureIds.empty(), ());
+                                   CHECK(!r2.m_featureIds.empty(), ());
+                                   if (r1.m_type != r2.m_type)
+                                     return r1.m_type < r2.m_type;
+                                   return r1.m_featureIds[0] < r2.m_featureIds[0];
+                                 });
+  auto const lower = range.first;
+  auto const upper = range.second;
+
+  // Checking if there's at least one restrictions of type Only from |featureIdFrom| and
+  // ending with |featureIdTo|. If yes, returns false.
+  if (lower != upper && lower->m_featureIds[1] == featureIdTo)
+    return false;
+
+  // Checking if there's at least one Only starting from |featureIdFrom|.
+  // If not, returns false.
+  if (lower == restrictions.cend() || lower->m_type != Restriction::Type::Only ||
+      lower->m_featureIds[0] != featureIdFrom)
+  {
+    return false;
+  }
+
+  // Note. At this point it's clear that there is an item in |restrictions| with type Only
+  // and starting with |featureIdFrom|. So there are two possibilities:
+  // * |edges| contains |it->m_featureIds[1]| => the end of |featureIdFrom| which implies in
+  //   |restrictions| is considered.
+  // * |edges| does not contain |it->m_featureIds[1]| => either the other end or an intermediate
+  // point of |featureIdFrom| is considered.
+  // See test FGraph_RestrictionF0F2Only for details.
+  CHECK_EQUAL(lower->m_featureIds.size(), 2, ("Only two link restrictions are support."));
+  for (SegmentEdge const & e : edges)
+  {
+    if (e.GetTarget().GetFeatureId() == lower->m_featureIds[isOutgoing ? 1 /* to */ : 0 /* from */])
```

Comment:
Как здесь написано - работает не верно. В общем, тут все несколько иначе должно выглядеть. Грубо, что-то в этом духе (см. ниже). Но вообще, наверно, лучше генерировать все restriction в тип No и во время исполнения только их обрабатывать.

```c++
void IndexGraph::GetEdgeList(Segment const & segment, bool isOutgoing, vector<SegmentEdge> & edges)
{
  RoadPoint const roadPoint = segment.GetRoadPoint(isOutgoing);
  Joint::Id const jointId = m_roadIndex.GetJointId(roadPoint);


  if (jointId != Joint::kInvalidId)
  {
    m_jointIndex.ForEachPoint(jointId, [&](RoadPoint const & rp) {
      GetNeighboringEdges(segment, rp, isOutgoing, edges);
    });
  }
  else
  {
    GetNeighboringEdges(segment, roadPoint, isOutgoing, edges);
  }
}

void IndexGraph::GetRestrictionEdgeList(Segment const & segment, bool isOutgoing,
                                        vector<SegmentEdge> & edges)
{
  vector<SegmentEdge> rawEdges;
  GetEdgeList(segment, isOutgoing, rawEdges);

  // Removing some edges according to restriction rules.
  edges.reserve(rawEdges.size());
  for (SegmentEdge const & e : rawEdges)
  {
    if (!IsRestricted(rawEdges, segment, e.GetTarget(), isOutgoing))
      edges.push_back(e);
  }
}

bool IndexGraph::IsRestricted(vector<SegmentEdge> const & edges, Segment const & u,
                              Segment const & v, bool isOutgoing)
{
  uint32_t const featureIdFrom = isOutgoing ? u.GetFeatureId() : v.GetFeatureId();
  uint32_t const featureIdTo = isOutgoing ? v.GetFeatureId() : u.GetFeatureId();

  // Looking for at least one restriction of type No from |featureIdFrom| to |featureIdTo|.
  if (binary_search(m_restrictions.cbegin(), m_restrictions.cend(),
                    Restriction(Restriction::Type::No, {featureIdFrom, featureIdTo})))
  {
    if (featureIdFrom != featureIdTo)
      return true;

    // @TODO(bykoianko) According to current code if a feature id is marked as a feature with
    // restrictricted U-turn it's restricted to make a U-turn on the both ends of the feature.
    // Generally speaking it's wrong. In osm there's information about the end of the feature
    // where the U-turn is restricted. It's necessary to pass the data to mwm and to use it here.
    // Please see test LineGraph_RestrictionF1F1No for details.
    return u.GetSegmentIdx() == v.GetSegmentIdx() && u.IsForward() != v.IsForward();
  }

  // Taking into account that number of restrictions of type Only starting and
  // ending with the same feature id is relevantly small it's possible to ignore such cases.
  if (featureIdFrom == featureIdTo)
    return false;

  // Checking if there's a restriction of type Only starting from |featureIdFrom| and
  // ending with |featureIdTo|. If yes, returns false.
  if (binary_search(m_restrictions.cbegin(), m_restrictions.cend(),
                    Restriction(Restriction::Type::Only, {featureIdFrom, featureIdTo})))
  {
    return false;
  }

  auto const typeAndFeatureFrom = [](Restriction const & r1, Restriction const & r2) {
    CHECK(!r1.m_featureIds.empty(), ());
    CHECK(!r2.m_featureIds.empty(), ());
    if (r1.m_type != r2.m_type)
      return r1.m_type < r2.m_type;
    return r1.m_featureIds[0] < r2.m_featureIds[0];
  };

  if (isOutgoing)
  {
    // Looking for a range of restrictins of type Only starting from |featureIdFrom|
    // and finishing with any feature.
    auto const range = equal_range(m_restrictions.cbegin(), m_restrictions.cend(),
                                   Restriction(Restriction::Type::Only, {featureIdFrom}),
                                   typeAndFeatureFrom);
    auto const lower = range.first;
    auto const upper = range.second;

    // Checking if there's a restriction of type Only starting from |featureIdFrom|. If not returns false.
    if (lower == upper)
      return false;

    // Note. The loop is necessary because it's possible that there's several restriction
    // of type Only starting from |featureIdFrom|.
    for (auto i = lower; i != upper; ++i)
    {
      CHECK_EQUAL(i->m_featureIds.size(), 2, ("Only two link restrictions are support."));

      // Note. At this point it's clear that there is an item in |restrictions| with type Only
      // and starting with |featureIdFrom|. So there are two possibilities:
      // * |edges| contains |it->m_featureIds[1]| => the end of |featureIdFrom| which implies in
      //   |restrictions| is considered.
      // * |edges| does not contain |it->m_featureIds[1]| => either the other end or an intermediate
      // point of |featureIdFrom| is considered.
      // See test FGraph_RestrictionF0F2Only for details.
      for (SegmentEdge const & e : edges)
      {
        if (e.GetTarget().GetFeatureId() == i->m_featureIds[1 /* to */])
          return true;
      }
    }
  }
  else // ingoing
  {
    // Checking each edge ingoing to segment |u|.
    for (SegmentEdge const & inEdge : edges)
    {
      // Looking for a range of restrictins of type Only starting from |inEdge.GetTarget().GetFeatureId()|
      // and finishing with any feature.
      auto const range = equal_range(m_restrictions.cbegin(), m_restrictions.cend(),
                                     Restriction(Restriction::Type::Only, {inEdge.GetTarget().GetFeatureId()}),
                                     typeAndFeatureFrom);
      auto const lower = range.first;
      auto const upper = range.second;

      // Checking if there's a restriction of type Only starting from |inEdge.GetTarget().GetFeatureId()|.
      // If not continue.
      if (lower == upper)
        continue;

      vector<SegmentEdge> outgoingFromInEdge;
      GetEdgeList(inEdge.GetTarget(), true /* isOutgoing */, outgoingFromInEdge);

      for (auto i = lower; i != upper; ++i)
      {
        CHECK_EQUAL(i->m_featureIds.size(), 2, ("Only two link restrictions are support."));

        // There are two possibilities at this point:
        // * there's a restriction starting from from feature id of |inEdge| and finishing at one of edge
        //   ingoing to |u| (|edges|). In that case it's impossible moving from |featureIdFrom|
        //   to |featureIdTo|.
        // * there's a restriction starting from from feature id of |inEdge| and finishing at an edge
        //   which is not connected with |featureIdTo|. In that case "continue" should be done.
        for (SegmentEdge const & e : outgoingFromInEdge)
        {
          if (featureIdTo == e.GetTarget().GetFeatureId())
            continue;
          if (i->m_featureIds[1 /* to */] == e.GetTarget().GetFeatureId())
            return true;
        }
      }
    }
  }
  return false;
}
```